format.js

```diff
@@ -0,0 +1,327 @@
+  SUB_FEED: 1,
+  END_OF_FEED: 2,
+}
+
+function newNativeMsg(opts) {
+  if (typeof opts.tag !== 'number') {
+    throw new Error('opts requires opts.tag, it must be a number')
+  }
+  if (opts.tag < 0 || opts.tag > 2) {
+    throw new Error('opts.tag must be either 0, 1 or 2')
+  }
+  const authorBFE = bfe.encode(opts.keys.id)
+  const previous = opts.previous || { key: null, value: { sequence: 0 } }
+  const previousBFE = bfe.encode(previous.key)
+  const contentBuffer = bipf.allocAndEncode(opts.content)
+  const contentHash = makeContentHash(contentBuffer)
+  const parentBFE = bfe.encode(opts.parent || null)
+  const tag = Buffer.from([opts.tag])
+  const sequence = previous.value.sequence + 1
+  const timestamp = +opts.timestamp
+
+  const value = [
+    authorBFE,
+    parentBFE,
+    sequence,
+    timestamp,
+    previousBFE,
+    tag,
+    contentBuffer.length,
+    contentHash,
+  ]
+
+  const encodedValue = bipf.allocAndEncode(value)
+  // FIXME: we need ssb-keys to support returning buffer from sign()
+  const signature = ssbKeys.sign(opts.keys, opts.hmacKey, encodedValue)
+  const sigBuf = _base64ToBuffer(signature)
+
+  return bipf.allocAndEncode([encodedValue, sigBuf, contentBuffer])
+}
+
+function _fromNativeMsgJS(nativeMsg) {
+  if (_jsMsgValCache.has(nativeMsg)) {
+    return _jsMsgValCache.get(nativeMsg)
+  }
+  const [encodedVal, sigBuf, contentBuf] = extract(nativeMsg)
+  const [
+    authorBFE,
+    parentBFE,
+    sequence,
+    timestamp,
+    previousBFE,
+    tag,
+    contentLength,
+    contentHashBuf,
+  ] = extractVal(encodedVal)
+  const author = bfe.decode(authorBFE)
+  const parent = bfe.decode(parentBFE)
+  const previous = bfe.decode(previousBFE)
+  const content = bipf.decode(contentBuf)
+  const contentHash = contentHashBuf
+  const signature = sigBuf
+  const msgVal = {
+    author,
+    parent,
+    sequence,
+    timestamp,
+    previous,
+    tag,
+    content,
+    contentHash,
+    signature,
+  }
+  _jsMsgValCache.set(nativeMsg, msgVal)
+  return msgVal
+}
+
+function _fromNativeMsgBIPF(nativeMsg) {
+  if (_bipfMsgValCache.has(nativeMsg)) {
+    return _bipfMsgValCache.get(nativeMsg)
+  }
+  const [encodedVal, sigBuf, contentBuf] = extract(nativeMsg)
+  const [
+    authorBFE,
+    parentBFE,
+    sequence,
+    timestamp,
+    previousBFE,
+    tag,
+    contentLength,
+    contentHash,
+  ] = extractVal(encodedVal)
+  const author = bfe.decode(authorBFE)
+  const parent = bfe.decode(parentBFE)
+  const previous = bfe.decode(previousBFE)
+  const signature = sigBuf
+  bipf.markIdempotent(contentBuf)
+  const msgVal = {
+    author,
+    parent,
+    sequence,
+    timestamp,
+    previous,
+    content: contentBuf,
+    contentHash,
+    signature,
+    tag,
+  }
+  const bipfMsg = bipf.allocAndEncode(msgVal)
+  _bipfMsgValCache.set(nativeMsg, bipfMsg)
+  return bipfMsg
+}
+
+function fromNativeMsg(nativeMsg, encoding = 'js') {
+  if (encoding === 'js') {
+    return _fromNativeMsgJS(nativeMsg)
```

Comment:
Could we call this `_fromNativeToMsgJS` instead. Same with `_fromNativeToMsgBIPF` ?

Reply:
Done